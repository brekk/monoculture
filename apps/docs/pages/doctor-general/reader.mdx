# reader



## Unknown

 @name iterateOverWorkspacesAndReadFiles  @future  @example  ```js test=true  import { fork, resolve as resolveF, parallel } from 'fluture'  // drgen-import-above  fork(done)(x => {    expect(x).toEqual([      "tools/digested",      "tools/doctor-general-cli",      "tools/gitparty",      "tools/spacework",      "tools/superorganism",      "tools/treacle",    ])    done()  })(    iterateOverWorkspacesAndReadFiles(      {        searchGlob: '*',        ignore: []      },      '../..',      resolveF(['tools/'])    )  )
