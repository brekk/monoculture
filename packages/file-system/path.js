import { join, normalize } from 'node:path'
import { curry } from 'ramda'

// pathJoin :: String -> String -> String
export const pathJoin = curry((pwd, x) => {
  // TODO: this should ostensibly be the province of a type system to ensure this safety
  if (typeof pwd !== 'string' || typeof x !== 'string') {
    throw new Error('Cannot normalize bad paths.')
  }
  return join(pwd, normalize(x))
})
