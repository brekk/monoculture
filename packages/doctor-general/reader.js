import { log } from './log'
import { chain, when, curry, flatten, map, pipe, propOr } from 'ramda'
import { parallel } from 'fluture'
import { signal } from 'kiddo'
import { readJSONFile, readDirWithConfig } from 'file-system'

export const iterateOverWorkspacesAndReadFiles = curry(
  function _iterateOverWorkspacesAndReadFiles(
    { search: searchGlob, ignore },
    root,
    x
  ) {
    log.reader('searchGlob?', searchGlob)
    return pipe(
      // look for specific file types
      map(workspace => workspace + searchGlob),
      log.core('ITERATING'),
      chain(
        readDirWithConfig({
          // exclude some search spaces
          ignore,
          cwd: root,
        })
      )
    )(x)
  }
)

export const readPackageJsonWorkspaces = curry(
  function _readPackageJsonWorkspaces(root, x) {
    return map(
      pipe(
        propOr([], 'workspaces'),
        log.parse('workspaces'),
        // we want directories only
        map(z => `${z}/`),
        map(readDirWithConfig({ cwd: root }))
      )
    )(x)
  }
)

export const monorepoRunner = curry(
  function _monorepoRunnerF(cancel, config, root, pkgJsonPath) {
    const { showMatchesOnly } = config
    log.core('CONFIG', config)
    return pipe(
      log.core('reading root package.json'),
      readJSONFile,
      readPackageJsonWorkspaces(root),
      chain(parallel(10)),
      signal(cancel, {
        text: 'Reading all workspaces...',
        successText: 'Read all workspaces!',
        failText: 'Unable to read all workspaces.',
      }),
      map(
        pipe(
          flatten,
          log.core('workflows, flat'),
          iterateOverWorkspacesAndReadFiles(config, root)
        )
      ),
      chain(parallel(10)),
      when(
        () => !showMatchesOnly,
        pipe(
          map(flatten),
          log.verbose('all files read'),
          signal(cancel, {
            text: 'Reading all files...',
            successText: 'Read all files!',
            failText: 'Unable to read all files.',
          })
        )
      )
    )(pkgJsonPath)
  }
)
