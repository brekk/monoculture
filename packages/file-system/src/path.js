import { join, normalize } from 'node:path'
import { curry } from 'ramda'

// pathRelativeTo :: String -> String -> String
export const pathRelativeTo = curry((pwd, x) => join(pwd, normalize(x)))
