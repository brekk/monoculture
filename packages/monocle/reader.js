import path from 'node:path'

import {
  identity as I,
  curry,
  pipe,
  map,
  trim,
  split,
  addIndex,
  chain,
} from 'ramda'
import { parallel } from 'fluture'

import { readDirWithConfig, readFile } from 'file-system'
import { futureFileProcessor } from 'monorail'

import { hash } from './hash'
import { log } from './log'

export const readMonoFile = curry(
  function _readMonoFile(basePath, trimContent, file) {
    return pipe(
      log.file('reading'),
      readFile,
      map(buf =>
        pipe(
          split('\n'),
          // ostensibly files have content that starts at a 1 index
          addIndex(map)((y, i) => [i + 1, trimContent ? trim(y) : y]),
          body => ({
            name: path.relative(basePath, file),
            hash: hash(buf),
            body,
          })
        )(buf)
      )
    )(file)
  }
)

export const readAll = curry(function _readAll(config, dirglob) {
  return pipe(
    log.file('reading glob'),
    readDirWithConfig({ ...config, nodir: true }),
    config.showMatchesOnly
      ? I
      : chain(files =>
          pipe(map(readMonoFile(config.cwd, config.trim)), parallel(10))(files)
        )
  )(dirglob)
})

export const monoprocessor = curry(
  function _monoprocessor(config, pluginsF, dirGlob) {
    return pipe(
      readAll(config),
      config.showMatchesOnly ? I : futureFileProcessor(config, pluginsF)
    )(dirGlob)
  }
)
