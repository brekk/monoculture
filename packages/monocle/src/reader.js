import path from 'node:path'

import { curry, pipe, map, split, addIndex, fromPairs, chain } from 'ramda'
import { parallel, resolve } from 'fluture'

import { readDirWithConfig, readFile } from 'file-system'
import { futureFileProcessor } from 'monorail'

import { hash } from './hash'

export const readMonoFile = curry((basePath, file) => {
  return pipe(
    readFile,
    map(buf =>
      pipe(
        split('\n'),
        addIndex(map)((y, i) => [i, y]),
        body => ({
          file: path.relative(basePath, file),
          hash: hash(buf),
          body,
        })
      )(buf)
    )
  )(file)
})

export const readAll = curry((config, dirglob) =>
  pipe(
    readDirWithConfig(config),
    chain(files =>
      pipe(map(readMonoFile(config.basePath)), parallel(10))(files)
    )
  )(dirglob)
)

export const monoprocessor = curry((config, plugins, dirGlob) =>
  pipe(readAll(config), futureFileProcessor(config, resolve(plugins)))(dirGlob)
)
