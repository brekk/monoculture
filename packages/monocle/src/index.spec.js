import path from 'node:path'
import { fork } from 'fluture'
import {
  toPairs,
  pipe,
  map,
  prop,
  trim,
  reduce,
  invertObj,
  filter,
} from 'ramda'
import { reader, monoprocessor } from './reader'

test('reader', done => {
  const files = reader(
    {
      ignore: [path.join(__dirname, '*.spec.js')],
      basePath: __dirname,
    },
    path.join(__dirname, '/*')
  )
  fork(e => {
    // eslint-disable-next-line no-console
    console.warn(e)
    done()
  })(
    pipe(z => {
      const out = map(prop('hash'))(z)
      expect(out).toEqual([
        '5cd99d9df3d8e4c2fcc83cf248314a4836c71d88f8d6e4fdc0d84c4e8c7b029d',
        'c58cb552a847bf29bec7e5c61d15cf07bbb9a000b8b0a565c7981c29ebf00f5a',
        '4e2ab6dbbaf93f2897db5ff8865837bd33850c7e6babaeb3c7c3bd20e488ba0c',
      ])
      done()
    })
  )(files)
})

const PLUGINS = [
  {
    name: 'wordcount',
    dependencies: [],
    fn: (c, f) =>
      pipe(
        map(([, lineContent]) => lineContent.split(' ').map(trim).length),
        reduce((a, b) => a + b, 0)
      )(f.body),
  },
  {
    name: 't-words',
    dependencies: [],
    preserveLine: true,
    fn: (c, line) => line.split(' ').filter(z => z.startsWith('t')).length,
  },
]

test('monoprocessor', done => {
  fork(done)(x => {
    const { state: rawState } = x
    const stateKeys = Object.keys(rawState)

    expect(stateKeys.includes('t-words')).toBeTruthy()
    expect(stateKeys.includes('wordcount')).toBeTruthy()

    expect(rawState.wordcount).toEqual([
      ['5cd99d9df3d8e4c2fcc83cf248314a4836c71d88f8d6e4fdc0d84c4e8c7b029d', 27],
      ['c58cb552a847bf29bec7e5c61d15cf07bbb9a000b8b0a565c7981c29ebf00f5a', 44],
      ['4e2ab6dbbaf93f2897db5ff8865837bd33850c7e6babaeb3c7c3bd20e488ba0c', 225],
    ])
    done()
  })(
    monoprocessor(
      {
        basePath: __dirname,
        nodir: true,
        // dot: true,
        ignore: [path.join(__dirname, '*.spec.js')],
      },
      PLUGINS,
      path.join(__dirname, '/**')
    )
  )
})
