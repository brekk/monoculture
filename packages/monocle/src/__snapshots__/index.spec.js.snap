// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`monoprocessor 1`] = `
{
  "c-words": {
    "5cd99d9df3d8e4c2fcc83cf248314a4836c71d88f8d6e4fdc0d84c4e8c7b029d": {
      "body": [
        [
          0,
          1,
        ],
        [
          1,
          0,
        ],
        [
          2,
          1,
        ],
        [
          3,
          2,
        ],
        [
          4,
          0,
        ],
        [
          5,
          0,
        ],
        [
          6,
          0,
        ],
        [
          7,
          0,
        ],
      ],
      "file": "hash.js",
      "hash": "5cd99d9df3d8e4c2fcc83cf248314a4836c71d88f8d6e4fdc0d84c4e8c7b029d",
    },
    "c87fe2ab04cdef09c054d7e5dd985cfcc97ad420e357226cb3e375c98434b619": {
      "body": [
        [
          0,
          0,
        ],
        [
          1,
          0,
        ],
        [
          2,
          2,
        ],
        [
          3,
          0,
        ],
        [
          4,
          0,
        ],
        [
          5,
          0,
        ],
        [
          6,
          0,
        ],
        [
          7,
          0,
        ],
        [
          8,
          0,
        ],
        [
          9,
          0,
        ],
        [
          10,
          2,
        ],
        [
          11,
          0,
        ],
        [
          12,
          0,
        ],
        [
          13,
          0,
        ],
        [
          14,
          0,
        ],
        [
          15,
          0,
        ],
        [
          16,
          0,
        ],
        [
          17,
          0,
        ],
        [
          18,
          0,
        ],
        [
          19,
          0,
        ],
        [
          20,
          0,
        ],
        [
          21,
          0,
        ],
        [
          22,
          0,
        ],
        [
          23,
          0,
        ],
        [
          24,
          0,
        ],
        [
          25,
          0,
        ],
        [
          26,
          0,
        ],
        [
          27,
          2,
        ],
        [
          28,
          0,
        ],
        [
          29,
          0,
        ],
        [
          30,
          1,
        ],
        [
          31,
          0,
        ],
        [
          32,
          0,
        ],
        [
          33,
          0,
        ],
        [
          34,
          0,
        ],
        [
          35,
          0,
        ],
        [
          36,
          2,
        ],
        [
          37,
          0,
        ],
        [
          38,
          0,
        ],
        [
          39,
          0,
        ],
      ],
      "file": "reader.js",
      "hash": "c87fe2ab04cdef09c054d7e5dd985cfcc97ad420e357226cb3e375c98434b619",
    },
    "f5ca9070a68f19933e3cd35a10e18013655b9b9a9c1b78ec2c7262a98c2fee17": {
      "body": [
        [
          0,
          0,
        ],
        [
          1,
          0,
        ],
        [
          2,
          0,
        ],
      ],
      "file": "index.js",
      "hash": "f5ca9070a68f19933e3cd35a10e18013655b9b9a9c1b78ec2c7262a98c2fee17",
    },
  },
  "wordcount": {
    "5cd99d9df3d8e4c2fcc83cf248314a4836c71d88f8d6e4fdc0d84c4e8c7b029d": 27,
    "c87fe2ab04cdef09c054d7e5dd985cfcc97ad420e357226cb3e375c98434b619": 225,
    "f5ca9070a68f19933e3cd35a10e18013655b9b9a9c1b78ec2c7262a98c2fee17": 9,
  },
}
`;

exports[`readAll 1`] = `
[
  "5cd99d9df3d8e4c2fcc83cf248314a4836c71d88f8d6e4fdc0d84c4e8c7b029d",
  "f5ca9070a68f19933e3cd35a10e18013655b9b9a9c1b78ec2c7262a98c2fee17",
  "c87fe2ab04cdef09c054d7e5dd985cfcc97ad420e357226cb3e375c98434b619",
]
`;
